### PR TITLE
Add Github Action to run TPU tests.

### PR DESCRIPTION
**Pull request by zcain117**
## What does this PR do?

Runs a Github Action to run TPU tests. For now the PR is set to run on incoming PRs but we may want to change to a post-submit hook due to constraints on Github Secrets.

Fixes # (issue)

# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you create a separate PR for every change.
- [ ] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests? 
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: .github/workflows/tpu-testing.yml [hunk 1]
```diff
@@ -0,0 +1,105 @@
+name: Run TPU tests on current PR.
```
Comment (Borda) [inline .github/workflows/tpu-testing.yml:1]:
```suggestion
name: TPU tests
```

File: .github/workflows/tpu-testing.yml [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+on:
+  pull_request:
+    branches:
+    - master
```
Comment (Borda) [inline .github/workflows/tpu-testing.yml:6]:
```suggestion
  push:
    branches:
      - master
  pull_request:
    branches:
      - master
```
it shall run also on master as the project stats are collected from amster

File: .github/workflows/tpu-testing.yml [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+
+env:
+  PROJECT_ID: ${{ secrets.GKE_PROJECT }}
+  GKE_CLUSTER: lightning-cluster   # TODO: update to cluster name.
+  GKE_ZONE: us-central1-a  # TODO: update to cluster zone.
```
Comment (Borda) [inline .github/workflows/tpu-testing.yml:11]: is the todo here still valid?

File: docker/tpu/tpu_example.jsonnet [hunk 4]
```diff
@@ -0,0 +1,33 @@
+local base = import 'templates/base.libsonnet';
```
Comment (Borda) [inline docker/tpu/tpu_example.jsonnet:1]: rather name it `tpu_test_case.jsonnet` then using missleading example

File: docker/tpu/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+RUN gsutil cp 'gs://tpu-pytorch/wheels/torch-nightly-cp36-cp36m-linux_x86_64.whl' .
+RUN gsutil cp 'gs://tpu-pytorch/wheels/torch_xla-nightly-cp36-cp36m-linux_x86_64.whl' .
+RUN gsutil cp 'gs://tpu-pytorch/wheels/torchvision-nightly-cp36-cp36m-linux_x86_64.whl' .
+RUN pip install 'torch-nightly-cp36-cp36m-linux_x86_64.whl'
+RUN pip install 'torch_xla-nightly-cp36-cp36m-linux_x86_64.whl'
+RUN pip install 'torchvision-nightly-cp36-cp36m-linux_x86_64.whl'
+RUN rm 'torch-nightly-cp36-cp36m-linux_x86_64.whl'
+RUN rm 'torch_xla-nightly-cp36-cp36m-linux_x86_64.whl'
+RUN rm 'torchvision-nightly-cp36-cp36m-linux_x86_64.whl'
+RUN apt-get install -y libomp5
```
Comment (Borda) [inline docker/tpu/Dockerfile:43]:
```suggestion
RUN pip uninstall -y torch && \
# TODO: Python 3.7 wheels are available. Replace cp36-cp36m with cp37-cp37m
    gsutil cp 'gs://tpu-pytorch/wheels/torch-nightly-cp36-cp36m-linux_x86_64.whl' . && \
    gsutil cp 'gs://tpu-pytorch/wheels/torch_xla-nightly-cp36-cp36m-linux_x86_64.whl' . && \
    gsutil cp 'gs://tpu-pytorch/wheels/torchvision-nightly-cp36-cp36m-linux_x86_64.whl' . && \
    pip install 'torch-nightly-cp36-cp36m-linux_x86_64.whl' && \
    pip install 'torch_xla-nightly-cp36-cp36m-linux_x86_64.whl' && \
    pip install 'torchvision-nightly-cp36-cp36m-linux_x86_64.whl' && \
    rm 'torch-nightly-cp36-cp36m-linux_x86_64.whl' && \
    rm 'torch_xla-nightly-cp36-cp36m-linux_x86_64.whl' && \
    rm 'torchvision-nightly-cp36-cp36m-linux_x86_64.whl' && \
    apt-get install -y libomp5
```

File: docker/tpu/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+    git checkout CI && \
+    cd .. && \
+    pip install ./pytorch-lightning
+
+# If using this image for tests, intall more dependencies and don't delete
+# the source code where the tests live.
+RUN if [ $TEST_IMAGE -eq 1 ] ; then pip install -r pytorch-lightning/requirements/test.txt ; else rm -rf pytorch-lightning ; fi
+
+RUN python -c "import pytorch_lightning as pl; print(pl.__version__)"
+RUN conda init bash
```
Comment (Borda) [inline docker/tpu/Dockerfile:60]: isn't it already on line L29?

**Comment by Borda** [comment]
this is the report https://codecov.io/gh/PyTorchLightning/pytorch-lightning/pull/2376 lets see if codecov merge both reports...

File: .github/workflows/tpu-testing.yml [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+        echo "Done waiting. Job status code: $status_code" && \
+        # Allow time for logs to flush.
+        sleep 60 && \
+        echo "JOB_NAME: $job_name" && \
+        echo "GKE_CLUSTER: $GKE_CLUSTER" && \
+        echo "GKE_ZONE: $GKE_ZONE" && \
+        gcloud logging read "resource.type=k8s_container resource.labels.project_id=$PROJECT_ID resource.labels.location=$GKE_ZONE resource.labels.cluster_name=$GKE_CLUSTER resource.labels.namespace_name=default resource.labels.pod_name:$job_name" --limit 10000000 --order asc --format 'value(textPayload)' --project=$PROJECT_ID > /tmp/full_output.txt && \
+        csplit /tmp/full_output.txt '/<?xml version="1.0" ?>/' && \
+        # First portion is the test logs. Print these to Github Action stdout.
+        cat xx00 && \
```
Comment (Borda) [inline .github/workflows/tpu-testing.yml:91]: later you upload `./xx01` so what is the relation?

Comment (zcain117) [reply .github/workflows/tpu-testing.yml:91]: xx00 is the first half of the `csplit` which contains test logs and xx01 is the second half which contains the coverage XML. These are the default filenames of the `csplit` command

File: docker/tpu/Dockerfile (hunk 6 above)
Comment (zcain117) [reply docker/tpu/Dockerfile:60]: This line modifies the `.bashrc` file. Last time I tried without it, the VM was not able to run `conda activate container` upon startup. I can try again at some point once I've made some more cleanups

File: .github/workflows/tpu-testing.yml (hunk 3 above)
Comment (zcain117) [reply .github/workflows/tpu-testing.yml:14]: no, this is done. These vars are working

File: docker/tpu/tpu_example.jsonnet (hunk 4 above)
Comment (zcain117) [reply docker/tpu/tpu_example.jsonnet]: I like `tpu_test_case.jsonnet`

right now I see `tpu_test-case.jsonnet` in your latest commits but I think 2 underscores is better than 1 underscore and 1 hyphen

Comment (Borda) [reply docker/tpu/tpu_example.jsonnet]: sure, let's do `tpu_test_case.jsonnet`

**Comment by Borda** [comment]
@zcain117 I have changed the used python to 3.7 and tested docker build locally, but then the TPU fails without any obvious reason... mind check it?

**Comment by Borda** [comment]
seems I found the issue, Conda does not forward the exit status... https://github.com/conda/conda/issues/8385
is there a reason to use Conda or can we use just PyPI?

File: docker/tpu/Dockerfile [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+    git checkout CI && \
+    cd .. && \
+    pip install ./pytorch-lightning
+
+# If using this image for tests, intall more dependencies and don't delete
+# the source code where the tests live.
+RUN if [ $TEST_IMAGE -eq 1 ] ; then \
+        pip install -r pytorch-lightning/requirements/test.txt ; \
+    else rm -rf pytorch-lightning ; \
+    fi
```
Comment (Borda) [inline docker/tpu/Dockerfile:62]:
```suggestion
    else \
        rm -rf pytorch-lightning ; \
    fi
```

**Comment by Borda** [comment]
seems that auto-scaling works fine, I think after passing the remaining test to is ready to go :tada: 
cc: @williamFalcon @PyTorchLightning/core-contributors

**Comment by williamFalcon** [comment]
amazing!! @zcain117 thank you so much :)

**Pull request merged**